File: all-submissions/1544-count-good-nodes-in-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int count = 0;
    
    public int goodNodes(TreeNode root) {
        if (root == null ) return 0;
        int maxVal = root.val;
        goodNodes(root.left, maxVal);
        goodNodes(root.right, maxVal);
        return count+1;
    }

    private void goodNodes(TreeNode root, int maxVal) {
        if(root == null) return;
        if (root.val >= maxVal) {
            maxVal = root.val;
            count++;
        } 
        goodNodes(root.left, maxVal); 
        goodNodes(root.right, maxVal);
    }
}

File: all-submissions/1544-count-good-nodes-in-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0
        res = 0
        max_val_so_far = root.val
        def dfs(node, max_val_so_far):
            if node is None:
                return 
            nonlocal res
            if node.val >= max_val_so_far:
                res += 1
                max_val_so_far = node.val
            dfs(node.left, max_val_so_far)
            dfs(node.right, max_val_so_far)
        dfs(root, max_val_so_far)
        return res
